### PR TITLE
point type abstraction layer

### PR DESCRIPTION
**Pull request by necipfazil**
- [x] Range abstraction
- [x] Point type abstraction

### CODE REVIEW

File: demos/PointAdapter/super4pcs_adapter_test.cc [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+          using MatcherType = gr::Match4pcsBase<gr::FunctorSuper4PCS, extlib1::PointAdapter, TrVisitorType, gr::AdaptivePointFilter, gr::AdaptivePointFilter::Options>;
+          using OptionType  = typename MatcherType::OptionsType;
+
+          OptionType options;
+          if(! Demo::setOptionsFromArgs(options, logger))
+          {
+            exit(-2); /// \FIXME use status codes for error reporting
+          }
+/* template<class PointType, class Range, class Matcher, class Options, class Sampler, class TransformVisitor> typename PointType::Scalar computeAlignment(const Options&, const gr::Utils::Logger&, const Range&, const Range&, Eigen::Ref<Eigen::Matrix<typename PointType::Scalar, 4, 4> >, const Sampler&, TransformVisitor&) */
+          extlib1::PointAdapter::Scalar score = computeAlignment<extlib1::PointAdapter, std::vector<extlib1::PointType1>, MatcherType, OptionType, SamplerType, TrVisitorType> (options, logger, set1_points1, set2_points1, mat, sampler, visitor);
```
Comment (nmellado) [inline demos/PointAdapter/super4pcs_adapter_test.cc]: You don't need to specify the template types if they can be deduced by the compiler: `Range`, `Options`, `Sampler` and `TransformVisitor`.
You may also need to reorder the template parameter to have all the deduced types on the right side of the others.

File: demos/PointAdapter/super4pcs_adapter_test.cc [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+#include "../demo-utils.h"
+
+#define sqr(x) ((x) * (x))
+
+using namespace std;
+using namespace gr;
+using namespace gr::Demo;
+
+// create a vector of extlib1::PointType1 from a vector of gr::Point3D 
+std::vector<extlib1::PointType1> getExtlib1Points(std::vector<gr::Point3D> grPoints)
```
[the remark this answers is not in the document]
Comment (necipfazil) [reply demos/PointAdapter/super4pcs_adapter_test.cc]: solved

File: demos/PointAdapter/super4pcs_adapter_test.cc (hunk 1 above)
Comment (necipfazil) [reply demos/PointAdapter/super4pcs_adapter_test.cc]: solved: https://github.com/necipfazil/OpenGR/commit/66bfc76354bc98837516d6460d6ab6ce30cc81e4

File: src/gr/shared.h [hunk 3]
```diff
@@ -152,6 +152,27 @@ class Point3D {
 //        Scalar overlap_estimation = 0.2;
 //    };
 
+// TODO: Document better
+// Wrapper around (any) PointType to allow mutation of position of PointType
+// through copying the internal position.
+template<typename PointType>
+struct PosMutablePoint : public PointType
```
Comment (nmellado) [inline src/gr/shared.h]: Hey,
Small question on this class: why not keeping it as internal class of `MatchBase` ? As it is used only internally to properly handle `Eigen::Map` with const access, we shouldn't need it outside of `MatchBase` and child classes, right ?

**Comment by necipfazil** [comment]
Design changes:
- Sampler fills sampled vector of pointers to input point cloud instead of directly copying the sampled points to out vector as in [sampling.h#L150](https://github.com/necipfazil/OpenGR/blob/c995fffdb79a564455d046b9b0660df0bc16ba2a/src/gr/sampling.h#L150). Rationalie: Matcher applies operations on the sampler's output vector that requires swap as in [matchBase.hpp#L316](https://github.com/necipfazil/OpenGR/blob/c995fffdb79a564455d046b9b0660df0bc16ba2a/src/gr/algorithms/matchBase.hpp#L316), which requires the instances to be move-constructible and move-assignable (does not work well with instances of PointAdapters that has members of type Eigen::Map). Having such operations on pointers of points relaxes this requirement. Wrapping of external point types are done after such operations as in [matchBase.hpp#L320](https://github.com/necipfazil/OpenGR/blob/c995fffdb79a564455d046b9b0660df0bc16ba2a/src/gr/algorithms/matchBase.hpp#L320).
- Traits::Coordinates is now a container of pointers to points instead of instances of points as in [https://github.com/necipfazil/OpenGR/blob/c995fffdb79a564455d046b9b0660df0bc16ba2a/src/gr/algorithms/match4pcsBase.h#L30](match4pcsBase.h#L30). Traits::Coordinates is not removed since it seems to provide a good view when coordinates are passed to methods as in [FunctorSuper4pcs.h#L51](https://github.com/necipfazil/OpenGR/blob/c995fffdb79a564455d046b9b0660df0bc16ba2a/src/gr/algorithms/FunctorSuper4pcs.h#L51). Could be changed such that Traits::Coordinates are no longer used, if needed.

Known issues, to be resolved:
- [ ] Super4PCS is not adapted to work with the abstraction layer. Taken out from cmakelist, to be adapted and added again.
- [ ] Tests are not adapted to work with the abstraction layer, which fails the release checks.
- [ ] Remove unused parameters from MatchBase::Initialize().

File: src/gr/shared.h (hunk 3 above)
Comment (necipfazil) [reply src/gr/shared.h]: Indeed.

I populate PosMutablePoint in point filters as in [PointPairFilter.h#L58](https://github.com/necipfazil/OpenGR/blob/c995fffdb79a564455d046b9b0660df0bc16ba2a/src/gr/algorithms/PointPairFilter.h#L58). However, point filter does not need to know it is dealing with PosMutablePoint instances either. I will refactor both (move PosMutablePoint to MatchBase, abstract point type in point filter without mentioning PosMutablePoint) accordingly.

Comment (necipfazil) [reply src/gr/shared.h]: solved https://github.com/STORM-IRIT/OpenGR/pull/31/commits/8fc03e9152df56cea0d507d7be5e16e22fee9149

PosMutablePoint is now a public member type of MatchBase class, propogated only to its children classes. Other classes, such as point filters, traits and functors are not explicity aware of this middle point type: they take it as template parameter.

The reasons why we can't make PosMutablePoint a protected member of MatchBase are the methods [MatchBase::getFirstSampled()](https://github.com/necipfazil/OpenGR/blob/8fc03e9152df56cea0d507d7be5e16e22fee9149/src/gr/algorithms/matchBase.h#L142) and [MatchBase::getSecondSampled()](https://github.com/necipfazil/OpenGR/blob/8fc03e9152df56cea0d507d7be5e16e22fee9149/src/gr/algorithms/matchBase.h#L147) (notice return type).

Comment (nmellado) [reply src/gr/shared.h]: Ok, I am fine with keeping `PosMutablePoint` public, as long as it is well documented.
It must be identified as a convenience class.

**Comment by necipfazil** [comment]
PointAdapter demo is removed. Now, Super4PCS can emulate PointAdapter usage. The type of point used can be changed by providing a command line argument ([-p point_type]).

- [ ] Tests will test the usage of different point types / adapters by running Super4PCS with different arguments

**Comment by nmellado** [comment]
@necipfazil thanks for the update.
Please update your branch with the latest updates from master to get the CI working on macOSx.

**Comment by necipfazil** [comment]
Known issue: with the templated IO methods that are moved from source files (io.cc) to header files (io.hpp), an external library (stb) is required at compile time for external apps that use OpenGR IO methods. One of the tests fail due to this issue. To be resolved.

**Pull request merged**